src/Raven.Server/Documents/Indexes/Static/JavaScriptReduceOperation.cs

```diff
@@ -43,48 +43,72 @@ private struct GroupBykeyComparer : IEqualityComparer<BlittableJsonReaderObject>
         {
             JavaScriptReduceOperation _parent;
             ReduceKeyProcessor _xKey, _yKey;
+            private BlittableJsonReaderObject _lastUsedBlittable;
+            private BlittableJsonReaderObject _lastUsedBucket;
+
             public GroupBykeyComparer(JavaScriptReduceOperation parent)
             {
                 _parent = parent;
                 _xKey = new ReduceKeyProcessor(_parent._groupByFields.Count(), CurrentIndexingScope.Current.UnmanagedBuffersPool);
                 _yKey = new ReduceKeyProcessor(_parent._groupByFields.Count(), CurrentIndexingScope.Current.UnmanagedBuffersPool);
                 _xKey.SetMode(ReduceKeyProcessor.Mode.MultipleValues);
                 _yKey.SetMode(ReduceKeyProcessor.Mode.MultipleValues);
+                _lastUsedBlittable = null;
+                _lastUsedBucket = null;
             }
 
             public unsafe bool Equals(BlittableJsonReaderObject x, BlittableJsonReaderObject y)
             {
-                _xKey.Reset();
-                _yKey.Reset();
+                var xCalCulated = ReferenceEquals(x, _lastUsedBucket);
```

Comment:
Too many branches on `Calculated`. 
Better to have this in a method and call it once.

Reply:
I'm not so sure it will have an impact, the bench prodiction is going to hit most of the time and the cost of iterating the fields twice is going to cost more.